### PR TITLE
[RF-DOCS]Active Record Associations Guide [ci-skip]

### PR DESCRIPTION
**Pull request by Ridhwana**
### Motivation / Background

Update the Active Record Association Documentation.

### Additional information

- [x] The images used to show the relationships of the different types of associations have both the table declarations, as well as a code sample, but the code sample is also in the guide. We could probably simplify the images by cutting off the code samples from them and leaving them in the guide code.
- [x] belongs_to section mentions bi-directional association and links to it within the same guide, but doesn't do the same for has_one and has_many, mentioned right after in the same paragraph.
- [x] There's a note about using bigint :supplier_id in the section between belongs_to and has_one, but up until that point it's always been using belongs_to in migrations... it's probably something that doesn't need to be in this area of the guide.
- [x] The has_many :through vs has_and_belongs_to_many section ends with a composite keys mention, we could link to the new guide there.
- [x] Associations between models with composite PKs may be simplified.
- [x] One example in bi-directional associations mentions Book belongs_to :writer, but then uses book.author a few lines below, which is incorrect.
- [x] STI section could be expanded, it's very useful in certain situations.
- [x] Add more info and make it clear
- [x] Group sections better
- [x] Amend the References section to be integrated with associations 
- [x] Abstract and remove duplications for options and scopes   

### Checklist

Before submitting the PR make sure the following are checked:

* [ ] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [ ] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [ ] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: guides/source/association_basics.md [hunk 1]
```diff
@@ -7,20 +7,56 @@ This guide covers the association features of Active Record.
...
-For example, consider a simple Rails application that includes a model for authors and a model for books. Each author can have many books.
+INFO: A macro-style call is a method that generates or modifies other methods at
+runtime, allowing for concise and expressive declarations of functionality, such
+as defining model associations in Rails. For example, `has_many :comments`.
 
-Without associations, the model declarations would look like this:
+When you set up an association, you instruct Rails to maintain the [Primary
+Key](https://en.wikipedia.org/wiki/Primary_key) and [Foreign
+Key](https://en.wikipedia.org/wiki/Foreign_key) information between instances of
+the two models, ensuring that your data stays consistent and properly linked.
```
Comment (rizwanreza) [inline guides/source/association_basics.md:33]: Perhaps pedantic but the foreign key and primary key information is maintained by the database that Rails can set up.

File: guides/source/association_basics.md [hunk 2]
```diff
@@ -7,20 +7,56 @@ This guide covers the association features of Active Record.
...
+```ruby
+class CreateAuthors < ActiveRecord::Migration[7.2]
+  def change
+    create_table :authors do |t|
+      t.string :name
+      t.timestamps
+    end
+
+    create_table :books do |t|
+      t.belongs_to :author
```
Comment (rizwanreza) [inline guides/source/association_basics.md:53]: You may want to use its alias `t.references` to avoid possible confusion with `belongs_to` in ActiveRecord classes.

File: guides/source/association_basics.md [hunk 3]
```diff
@@ -128,10 +207,27 @@ class CreateBooks < ActiveRecord::Migration[7.2]
...
-When used alone, `belongs_to` produces a one-directional one-to-one connection. Therefore each book in the above example "knows" its author, but the authors don't know about their books.
-To setup a [bi-directional association](#bi-directional-associations) - use `belongs_to` in combination with a `has_one` or `has_many` on the other model, in this case the Author model.
+When used alone, `belongs_to` produces a one-directional one-to-one connection.
+Therefore each book in the above example "knows" its author, but the authors
+don't know about their books. To setup a [bi-directional
+association](#bi-directional-associations) - use `belongs_to` in combination
+with a `has_one` or `has_many` on the other model, in this case the Author
+model.
+
+NOTE: By default `belongs_to` validates the presence of the association to
```
Comment (rizwanreza) [inline guides/source/association_basics.md:217]:
```suggestion
NOTE: By default `belongs_to` validates the presence of the associated record to
```

File: guides/source/association_basics.md [hunk 4]
```diff
@@ -355,2288 +439,2469 @@ class CreateAccountHistories < ActiveRecord::Migration[7.2]
...
 ```ruby
-@book_ids = @author.book_ids
+manager = employee.manager
 ```
 
-##### `collection_singular_ids=(ids)`
+[connection.add_reference]:
+    https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/SchemaStatements.html#method-i-add_reference
+
+Single Table Inheritance (STI)
```
Comment (mstroming) [inline guides/source/association_basics.md:1863]: I think it would be useful to mention that STI breaks referential integrity.

File: guides/source/association_basics.md [hunk 5]
```diff
@@ -355,2288 +439,2469 @@ class CreateAccountHistories < ActiveRecord::Migration[7.2]
...
+Delegated Types
+----------------
 
-If you set the `:autosave` option to `true`, Rails will save any loaded association members and destroy members that are marked for destruction whenever you save the parent object. Setting `:autosave` to `false` is not the same as not setting the `:autosave` option. If the `:autosave` option is not present, then new associated objects will be saved, but updated associated objects will not be saved.
+[`Single Table Inheritance (STI)`](#single-table-inheritance-sti) works best
+when there is little difference between subclasses and their attributes, but it
+includes all attributes of all subclasses in a single table.
 
-##### `:class_name`
+The disadvantage of this approach is that it can result in table bloat, as the
```
Comment (mstroming) [inline guides/source/association_basics.md:2038]: It would be helpful to mention the lack of referential integrity as another disadvantage.

**Review by mstroming — COMMENTED** [review]
For certain patterns, like STI and Delegated Types, I think it would be helpful to acknowledge the lack of referential integrity as a downside to these approaches.

File: guides/source/association_basics.md (hunk 2 above)
Comment (MatheusRich) [reply guides/source/association_basics.md:53]: Should we mention somewhere that these are aliases?

Comment (MatheusRich) [reply guides/source/association_basics.md:53]: Also, whatever we pick, let's keep it consistent throughout the page.

File: guides/source/association_basics.md [hunk 6]
```diff
@@ -140,210 +236,198 @@ create_table :books do |t|
 end
 ```
 
-### The `has_one` Association
+This ensures that even if `optional: true` allows `author_id` to be NULL, when
+it is not NULL, it must point to a valid record in the authors table.
```
Comment (MatheusRich) [inline guides/source/association_basics.md:240]: I found this sentence a bit hard to parse at first. Here's an attempt at a rewrite:

```suggestion
This ensures that even if `optional: true` is set, allowing `author_id` to be NULL, when
it the value is not NULL, it must point to a valid record in the authors table.
```

File: guides/source/association_basics.md [hunk 7]
```diff
@@ -140,210 +236,198 @@ create_table :books do |t|
...
 
-A [`has_one`][] association indicates that one other model has a reference to this model. That model can be fetched through this association.
+#### Methods Added by `belongs_to`
 
-For example, if each supplier in your application has only one account, you'd declare the supplier model like this:
+In database terms, the `belongs_to` association says that this model's table
+contains a column which represents a reference to another table. This can be
+used to set up one-to-one or one-to-many relations, depending on the setup. If
+the table of the other class contains the reference in a one-to-one relation,
+then you should use `has_one` instead.
```
Comment (MatheusRich) [inline guides/source/association_basics.md:248]:
```suggestion
used to set up one-to-one or one-to-many relations, depending on the setup. Howerver, if
the table _of the other class_ is the one containing the reference in a one-to-one relation,
then you should use `has_one` instead.
```

Comment (MatheusRich) [reply guides/source/association_basics.md:248]: I like this clarification about the difference between `has_one` and `belongs_to` (a classic beginner question), but it feels out of place in this subsection.

File: guides/source/association_basics.md [hunk 8]
```diff
@@ -140,210 +236,198 @@ create_table :books do |t|
...
 
-```ruby
-class Supplier < ApplicationRecord
-  has_one :account
-end
-```
+When you declare a `belongs_to` association, the declaring class automatically
+gains numerous methods related to the association. Some of these include:
```
Comment (MatheusRich) [inline guides/source/association_basics.md:251]:
```suggestion
gains numerous methods related to the association. Some of these are:
```

File: guides/source/association_basics.md [hunk 9]
```diff
@@ -140,210 +236,198 @@ create_table :books do |t|
...
+end
+```
+
+```irb
+irb> raise_validation_error: Validation failed: Name can't be blank (ActiveRecord::RecordInvalid)
 ```
 
-Then new join models are automatically created for the newly associated objects.
-If some that existed previously are now missing, then their join rows are automatically deleted.
+##### Checking for Association Changes
```
Comment (MatheusRich) [inline guides/source/association_basics.md:370]: Do you think it's worth mentioning that `model.association_changed?` and `model.association.changed?` are different? The first one tracks if there's a new record, while the second one tracks if some attribute in the association changed. If we add this note, _maybe_ it is worth also mentioning that this is backed by `ActiveModel::Dirty`

File: guides/source/association_basics.md (hunk 2 above)
Comment (Ridhwana) [reply guides/source/association_basics.md]: > Should we mention somewhere that these are aliases?

Since this is related to Active Record Migrations, folks will refer to that guide to learn more about migrations and [we mention that `belongs_to` is an alias there. ](https://guides.rubyonrails.org/active_record_migrations.html#creating-associations)

Comment (Ridhwana) [reply guides/source/association_basics.md]: > You may want to use its alias t.references to avoid possible confusion with belongs_to in ActiveRecord classes.

I think that the `belongs_to` alias is explicit and makes the migrations easier to read, however, I see your point about possible confusion here and so I've changed it to references but I hope that it's not misleading. 

> Also, whatever we pick, let's keep it consistent throughout the page.

@MatheusRich - my ears perk up when folks say we should keep it consistent because I agree wholeheartedly! In this case, I've used `t.belongs_to` in all the scenarios except the one that @rizwanreza pointed out because even though we can use the alias in this case, I agree that it may cause confusion.

File: guides/source/association_basics.md (hunk 5 above)
Comment (Ridhwana) [reply guides/source/association_basics.md]: @mstroming thanks - I've added a [considerations section](https://github.com/rails/rails/blob/2a8a870deb4b79cb310619fdc304d89e9e462a3d/guides/source/association_basics.md#considerations) to STI to maintain referential integrity.

File: guides/source/association_basics.md (hunk 6 above)
Comment (Ridhwana) [reply guides/source/association_basics.md:242]: Thanks @MatheusRich, there was a typo in the suggested text, but I used a variation of it to make it easier to read ✨

File: guides/source/association_basics.md (hunk 7 above)
Comment (Ridhwana) [reply guides/source/association_basics.md:250]: Nice catch! I agree, I've gone through all the sections and moved the opening paragraph to after I introduce the migration for the association :)

File: guides/source/association_basics.md (hunk 8 above)
Comment (Ridhwana) [reply guides/source/association_basics.md:253]: Updated this sentence for all of the association method sections, better readability -thanks!

File: guides/source/association_basics.md [hunk 10]
```diff
@@ -355,2288 +439,2469 @@ class CreateAccountHistories < ActiveRecord::Migration[7.2]
...
+statement returns `false` and the assignment itself is cancelled.
 
-Similarly, you can retrieve `@product.pictures`.
+If the parent object (the one declaring the `has_one` association) is unsaved
+(that is, `new_record?` returns `true`) then the child objects are not saved.
+They will automatically when the parent object is saved.
 
-If you have an instance of the `Picture` model, you can get to its parent via `@picture.imageable`. To make this work, you need to declare both a foreign key column and a type column in the model that declares the polymorphic interface:
+If you want to assign an object to a `has_one` association without saving the
+object, use the `build_association` method.
```
[the remark this answers is not in the document]
Comment (Ridhwana) [reply guides/source/association_basics.md:617]: Yeah, I think it's useful - done  ☑️

File: guides/source/association_basics.md (hunk 9 above)
Comment (Ridhwana) [reply guides/source/association_basics.md:372]: Sounds good! I've added a note to the bottom:

```
NOTE: Do not confuse `model.association_changed?` with
`model.association.changed?`. The former checks if the association has been
replaced with a new record, while the latter tracks changes to the attributes of
the association.
```

However, I've chosen not to mention `ActiveModel::Dirty` as I think its more relevant to the ActiveModel section and we don't need to go into details here.

File: guides/source/association_basics.md [hunk 11]
```diff
@@ -355,2288 +447,2474 @@ class CreateAccountHistories < ActiveRecord::Migration[7.2]
...
-##### `order`
+WARNING: The precedence between model names is calculated using the `<=>`
+operator for `String`. This means that if the strings are of different lengths,
+and the strings are equal when compared up to the shortest length, then the
+longer string is considered of higher lexical precedence than the shorter one.
+For example, one would expect the tables "paper_boxes" and "papers" to generate
+a join table name of "papers_paper_boxes" because of the length of the name
+"paper_boxes", but it in fact generates a join table name of
+"paper_boxes_papers" (because the underscore '\_' is lexicographically _less_
+than 's' in common encodings).
```
Comment (p8) [inline guides/source/association_basics.md:2351]: This warning probably makes more sense in the migrations guide.

File: guides/source/association_basics.md [hunk 12]
```diff
@@ -355,2288 +447,2474 @@ class CreateAccountHistories < ActiveRecord::Migration[7.2]
...
-##### `collection.create!(attributes = {})`
+When you execute `@user.todos.create`, the `@todo` record will have its
+`user_id` value set to the `guid` value of `@user`.
 
-Does the same as `collection.create`, but raises `ActiveRecord::RecordInvalid` if the record is invalid.
+`has_and_belongs_to_many` does not support the `:primary_key` option. For this
+type of association, you can achieve similar functionality by using a join table
+with has_many `:through` association, which gives more flexibility and supports
+the `:primary_key` option. You can read more about this in the [`has_many
+:through` section](#has-many-through).
```
Comment (p8) [inline guides/source/association_basics.md:2825]: This is formatted a bit strangely. Not sure how to fix it:
image

File: guides/source/association_basics.md [hunk 13]
```diff
@@ -355,2288 +447,2474 @@ class CreateAccountHistories < ActiveRecord::Migration[7.2]
...
+
+The `collection` method returns a Relation of all of the associated objects. If
+there are no associated objects, it returns an empty Relation.
+
+```ruby
+@books = @author.books
 ```
 
-### Self Joins
+##### `collection<<(object, ...)`
```
Comment (p8) [inline guides/source/association_basics.md:806]: Could we group some of these methods by their use-case? Similar to `belongs_to` and `has_one`?

##### Querying the Association
    collection
    collection_singular_ids
    collection_singular_ids=(ids)
    collection.empty?
    collection.size
    collection.find(...)
    collection.where(...)
    collection.exists?(...)

##### Modifying the Association
    collection<<(object, ...)
    collection=(objects)
    collection.build(attributes = {})
    collection.create(attributes = {})
    collection.create!(attributes = {})
    collection.delete(object, ...)
    collection.destroy(object, ...)
    collection.clear
    collection.reload

Same thing applies to the `has_many :through` association.

File: guides/source/association_basics.md [hunk 14]
```diff
@@ -46,7 +87,12 @@ end
 @author.destroy
 ```
 
-With Active Record associations, we can streamline these - and other - operations by declaratively telling Rails that there is a connection between the two models. Here's the revised code for setting up authors and books:
+### Using Associations
+
+However, with associations, we can streamline these operations, as well as
+others, by explicitly informing Rails about the connection between the two
```
[the remark this answers is not in the document]
Comment (Ridhwana) [reply guides/source/association_basics.md:93]: Makes sense! I've updated the necessary instances. I've only kept those that reference the API doc.

File: guides/source/association_basics.md [hunk 15]
```diff
@@ -355,2288 +447,2474 @@ class CreateAccountHistories < ActiveRecord::Migration[7.2]
...
+      t.timestamps
+    end
+  end
+end
+```
 
-#### Methods Added by `has_many`
+Since polymorphic associations rely on storing class names in the database, that
+data must remain synchronized with the class name used by the Ruby code. When
+renaming a class, make sure to update the data in the polymorphic type column.
```
[the remark this answers is not in the document]
Comment (Ridhwana) [reply guides/source/association_basics.md:1704]: Yes, I think so, thanks!

File: guides/source/association_basics.md [hunk 16]
```diff
@@ -355,2288 +447,2474 @@ class CreateAccountHistories < ActiveRecord::Migration[7.2]
...
-##### `:extend`
+### Setting up Delegated Types
 
-Specifies a module or array of modules that will be extended into the association object returned. Useful for defining methods on associations, especially when they should be shared between multiple association objects.
+To use delegated types, we need to model our data as follows:
 
-##### `:foreign_key`
+* There is a superclass that stores shared attributes among all subclasses in
+  its table.
+* Each subclass must inherit from the super class, and will have a separate
```
[the remark this answers is not in the document]
Comment (Ridhwana) [reply guides/source/association_basics.md:2080]: Really good catch, it seems that the correct terms are "superclass" and "subclass". Will update all occurrences!

File: guides/source/association_basics.md [hunk 17]
```diff
@@ -355,2288 +447,2474 @@ class CreateAccountHistories < ActiveRecord::Migration[7.2]
...
+  def change
+    add_reference :books, :author
+  end
 end
 ```
 
-##### `offset`
 
-The `offset` method lets you specify the starting offset for fetching objects via an association. For example, `-> { offset(11) }` will skip the first 11 records.
+#### Creating Join Tables for `has_and_belongs_to_many` Associations
```
[the remark this answers is not in the document]
Comment (Ridhwana) [reply guides/source/association_basics.md:2334]: I like that we have this section in the association guides because it's very specific to the relationship and makes it clear that you need the right schema. It’s also easy to see what needs to be done, and everything is in one place. I added a link to the [Active Record Migrations Guides](https://edgeguides.rubyonrails.org/active_record_migrations.html#creating-a-join-table) so people can dive deeper into creating a join table. But I’d like to keep this section here too, unless you feel really strongly about removing it, then I can?

File: guides/source/association_basics.md (hunk 11 above)
Comment (Ridhwana) [reply guides/source/association_basics.md:2351]: Agreed its too much detail for here - Moved it to the relevant section in migrations.

File: guides/source/association_basics.md [hunk 18]
```diff
@@ -355,2288 +447,2474 @@ class CreateAccountHistories < ActiveRecord::Migration[7.2]
...
-* [`collection.empty?`][]
-* [`collection.size`][]
-* [`collection.find(...)`][`collection.find`]
-* [`collection.where(...)`][`collection.where`]
-* [`collection.exists?(...)`][`collection.exists?`]
-* [`collection.build(attributes = {})`][`collection.build`]
-* [`collection.create(attributes = {})`][`collection.create`]
-* [`collection.create!(attributes = {})`][`collection.create!`]
-* [`collection.reload`][]
+### Models with Composite Primary Keys
```
[the remark this answers is not in the document]
Comment (Ridhwana) [reply guides/source/association_basics.md:1712]: Done, since I spent time on the "Models with Composite Primary Keys", I updated the section in [active_record_composite_primary_keys#associations-between-models-with-composite-primary-keys](https://249b34bc.rails-docs-preview.pages.dev/guides/active_record_composite_primary_keys#associations-between-models-with-composite-primary-keys) with the changes and I referenced the section from the associations guide.

File: guides/source/association_basics.md (hunk 12 above)
Comment (Ridhwana) [reply guides/source/association_basics.md:2825]: Fixed 3 instances ✅
Screenshot 2024-09-01 at 22 41 56

File: guides/source/association_basics.md (hunk 13 above)
Comment (Ridhwana) [reply guides/source/association_basics.md]: ✅

File: guides/source/association_basics.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (p8) [reply guides/source/association_basics.md:2200]: That makes sense. If we are keeping this section, I think the following warning about model precedence in table naming should probably be moved to the migration guide.
I think it's too detailed for this section and it isn't mentioned in the migrations guide.

The warning is a bit unclear as well with the lexical scope ordering, while we mention alphabetical ordering elsewhere.

[the remark this answers is not in the document]
Comment (Ridhwana) [reply guides/source/association_basics.md:2200]: Totally agree. I had [moved the warning to the migrations docs yesterday and changed the term in active record migrations to lexical which seems more accurate based on the warning](https://github.com/rails/rails/pull/52523/commits/4d45a7e593f2a88167b2efe99973063c4575eca1) but I'd missed removing the warning from here, which was my intention. Removing it now.

[the remark this answers is not in the document]
Comment (Ridhwana) [reply guides/source/association_basics.md:2200]: Oh wait, no I did actually remove it already ( I was looking at my active record validations branch), here's the [commit](https://github.com/rails/rails/pull/52523/commits/4d45a7e593f2a88167b2efe99973063c4575eca1)

[the remark this answers is not in the document]
Comment (p8) [reply guides/source/association_basics.md:2200]: I've created a PR the add more sorting examples the the API docs:
https://github.com/rails/rails/pull/52766

[the remark this answers is not in the document]
Comment (p8) [reply guides/source/association_basics.md:2200]: Ah nevermind. You already made the changes 🙃

File: guides/source/association_basics.md (hunk 12 above)
Comment (p8) [reply guides/source/association_basics.md]: Nice!

File: guides/source/association_basics.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (p8) [reply guides/source/association_basics.md:2200]: This is looking great! I'll have a last quick scan through the whole guide later today and I'll merge it.

[the remark this answers is not in the document]
Comment (Ridhwana) [reply guides/source/association_basics.md:2200]: Thank you so much!

File: guides/source/association_basics.md [hunk 19]
```diff
@@ -7,20 +7,58 @@ This guide covers the association features of Active Record.
...
 
-Why Associations?
+Associations Overview
 -----------------
 
-In Rails, an _association_ is a connection between two Active Record models. Why do we need associations between models? Because they make common operations simpler and easier in your code.
+_Associations_ in Rails are special macro style calls that make it easy to
+connect different models. You can use them to tell Rails how your models relate
+to each other, which helps you manage your data more effectively, and makes
+common operations simpler and easier to read.
```
Comment (p8) [inline guides/source/association_basics.md:24]: Final nitpick...

The Migrations, Callbacks and Active Record Basics guides talk about "Active Record Migration", "Active Record Callbacks" and "Active Record Associations" instead of "Rails Associations". Rails doesn't support other types of associations but would it be clearer to stick to "Active Record Associations" here as well?

Also can we avoid "macro style calls" in the opening sentence? 😄

Maybe something like:
```suggestion
Active Record associations allow you to define relationships between models.
_Associations_ are implemented as special macro style calls that make it easy to tell Rails how your models relate
to each other, which helps you manage your data more effectively, and makes
common operations simpler and easier to read.
```

File: guides/source/active_record_composite_primary_keys.md [hunk 20]
```diff
@@ -123,9 +123,14 @@ guide to learn more.
 Associations between Models with Composite Primary Keys
```
Comment (p8) [inline guides/source/active_record_composite_primary_keys.md:123]: This section rewrite is a big improvement!
